File: libs/php/tests/CreditCardValidatorTest.php
<?php

use PHPUnit\Framework\TestCase;
use CreditCard\Identifier\CreditCardValidator;

require_once __DIR__ . '/../src/BrandData.php';
require_once __DIR__ . '/../src/BrandDataDetailed.php';
require_once __DIR__ . '/../src/CreditCardValidator.php';

class CreditCardValidatorTest extends TestCase
{
    private $validator;
    
    protected function setUp(): void
    {
        $this->validator = new CreditCardValidator();
    }
    
    public function testLuhnValidNumbers()
    {
        $this->assertTrue(CreditCardValidator::luhn('4012001037141112'));
        $this->assertTrue(CreditCardValidator::luhn('5533798818319497'));
        $this->assertTrue(CreditCardValidator::luhn('378282246310005'));
    }
    
    public function testLuhnInvalidNumbers()
    {
        $this->assertFalse(CreditCardValidator::luhn('1234567890123456'));
        $this->assertFalse(CreditCardValidator::luhn(''));
        $this->assertFalse(CreditCardValidator::luhn('4012001037141113'));
    }
    
    public function testFindBrandVisa()
    {
        $this->assertEquals('visa', $this->validator->findBrand('4012001037141112'));
        $this->assertEquals('visa', $this->validator->findBrand('4551870000000183'));
        $this->assertEquals('visa', $this->validator->findBrand('6367000000001022'));
    }
    
    public function testFindBrandMastercard()
    {
        $this->assertEquals('mastercard', $this->validator->findBrand('5533798818319497'));
        $this->assertEquals('mastercard', $this->validator->findBrand('5437251265160938'));
        $this->assertEquals('mastercard', $this->validator->findBrand('2221000000000000'));
    }
    
    public function testFindBrandAmex()
    {
        $this->assertEquals('amex', $this->validator->findBrand('378282246310005'));
        $this->assertEquals('amex', $this->validator->findBrand('376411112222331'));
        $this->assertEquals('amex', $this->validator->findBrand('371449635398431'));
    }
    
    public function testFindBrandDiscover()
    {
        $this->assertEquals('discover', $this->validator->findBrand('6011236044609927'));
        $this->assertEquals('discover', $this->validator->findBrand('6011091915358231'));
    }
    
    public function testFindBrandDiners()
    {
        $this->assertEquals('diners', $this->validator->findBrand('30066909048113'));
        $this->assertEquals('diners', $this->validator->findBrand('30266056449987'));
        $this->assertEquals('diners', $this->validator->findBrand('36490102462661'));
    }
    
    public function testFindBrandElo()
    {
        $this->assertEquals('elo', $this->validator->findBrand('6362970000457013'));
        $this->assertEquals('elo', $this->validator->findBrand('6363680000000000'));
    }
    
    public function testFindBrandHipercard()
    {
        $this->assertEquals('hipercard', $this->validator->findBrand('6062825624254001'));
        $this->assertEquals('hipercard', $this->validator->findBrand('6062821294950895'));
    }
    
    public function testFindBrandUnsupported()
    {
        $this->assertNull($this->validator->findBrand('1234567890123456'));
        $this->assertNull($this->validator->findBrand(''));
    }
    
    public function testIsSupported()
    {
        $this->assertTrue($this->validator->isSupported('4012001037141112'));
        $this->assertTrue($this->validator->isSupported('5533798818319497'));
        $this->assertFalse($this->validator->isSupported('1234567890123456'));
        $this->assertFalse($this->validator->isSupported(''));
    }
    
    public function testValidateCvv()
    {
        $this->assertTrue($this->validator->validateCvv('123', 'visa'));
        $this->assertTrue($this->validator->validateCvv('1234', 'amex'));
        $this->assertFalse($this->validator->validateCvv('12', 'visa'));
        $this->assertFalse($this->validator->validateCvv('1234', 'visa'));
        $this->assertFalse($this->validator->validateCvv('', 'visa'));
        $this->assertFalse($this->validator->validateCvv('123', 'unknown'));
    }
    
    public function testGetBrandInfo()
    {
        $visaInfo = $this->validator->getBrandInfo('visa');
        $this->assertNotNull($visaInfo);
        $this->assertEquals('visa', $visaInfo['name']);
        
        $this->assertNull($this->validator->getBrandInfo('unknown'));
    }
    
    public function testGetBrandInfoDetailed()
    {
        $visaDetailed = $this->validator->getBrandInfoDetailed('visa');
        $this->assertNotNull($visaDetailed);
        $this->assertEquals('visa', $visaDetailed['scheme']);
        $this->assertEquals('Visa', $visaDetailed['brand']);
        $this->assertNotEmpty($visaDetailed['number']['lengths']);
        
        $this->assertNull($this->validator->getBrandInfoDetailed('unknown'));
    }
    
    public function testFindBrandDetailed()
    {
        $brand = $this->validator->findBrandDetailed('4012001037141112');
        $this->assertNotNull($brand);
        $this->assertEquals('visa', $brand['scheme']);
        $this->assertEquals('Visa', $brand['brand']);
    }
    
    public function testListBrands()
    {
        $brands = $this->validator->listBrands();
        $this->assertNotEmpty($brands);
        $this->assertContains('visa', $brands);
        $this->assertContains('mastercard', $brands);
    }
}

File: libs/java/src/test/java/br/com/s2n/creditcard/identifier/CreditCardValidatorTest.java
package br.com.s2n.creditcard.identifier;

import org.junit.Test;
import static org.junit.Assert.*;

/**
 * Tests for CreditCardValidator.
 */
public class CreditCardValidatorTest {
    
    private final CreditCardValidator validator = new CreditCardValidator();
    
    @Test
    public void testLuhnValidNumbers() {
        assertTrue("Valid Visa", CreditCardValidator.luhn("4012001037141112"));
        assertTrue("Valid Mastercard", CreditCardValidator.luhn("5533798818319497"));
        assertTrue("Valid Amex", CreditCardValidator.luhn("378282246310005"));
    }
    
    @Test
    public void testLuhnInvalidNumbers() {
        assertFalse("Invalid number", CreditCardValidator.luhn("1234567890123456"));
        assertFalse("Empty string", CreditCardValidator.luhn(""));
        assertFalse("Modified valid number", CreditCardValidator.luhn("4012001037141113"));
    }
    
    @Test
    public void testFindBrandVisa() {
        assertEquals("visa", validator.findBrand("4012001037141112"));
        assertEquals("visa", validator.findBrand("4551870000000183"));
        assertEquals("visa", validator.findBrand("6367000000001022"));
    }
    
    @Test
    public void testFindBrandMastercard() {
        assertEquals("mastercard", validator.findBrand("5533798818319497"));
        assertEquals("mastercard", validator.findBrand("5437251265160938"));
        assertEquals("mastercard", validator.findBrand("2221000000000000"));
    }
    
    @Test
    public void testFindBrandAmex() {
        assertEquals("amex", validator.findBrand("378282246310005"));
        assertEquals("amex", validator.findBrand("376411112222331"));
        assertEquals("amex", validator.findBrand("371449635398431"));
    }
    
    @Test
    public void testFindBrandDiscover() {
        assertEquals("discover", validator.findBrand("6011236044609927"));
        assertEquals("discover", validator.findBrand("6011091915358231"));
    }
    
    @Test
    public void testFindBrandDiners() {
        assertEquals("diners", validator.findBrand("30066909048113"));
        assertEquals("diners", validator.findBrand("30266056449987"));
        assertEquals("diners", validator.findBrand("36490102462661"));
    }
    
    @Test
    public void testFindBrandElo() {
        assertEquals("elo", validator.findBrand("6362970000457013"));
        assertEquals("elo", validator.findBrand("6363680000000000"));
        assertEquals("elo", validator.findBrand("6277800000000000"));
    }
    
    @Test
    public void testFindBrandHipercard() {
        assertEquals("hipercard", validator.findBrand("6062825624254001"));
        assertEquals("hipercard", validator.findBrand("6062821294950895"));
    }
    
    @Test
    public void testFindBrandUnsupported() {
        assertNull("Unsupported card", validator.findBrand("1234567890123456"));
        assertNull("Empty string", validator.findBrand(""));
    }
    
    @Test
    public void testIsSupported() {
        assertTrue("Visa supported", validator.isSupported("4012001037141112"));
        assertTrue("Mastercard supported", validator.isSupported("5533798818319497"));
        assertFalse("Unsupported card", validator.isSupported("1234567890123456"));
        assertFalse("Empty string", validator.isSupported(""));
    }
    
    @Test
    public void testValidateCvv() {
        assertTrue("Visa CVV 123", validator.validateCvv("123", "visa"));
        assertTrue("Amex CVV 1234", validator.validateCvv("1234", "amex"));
        assertFalse("Visa CVV 12", validator.validateCvv("12", "visa"));
        assertFalse("Visa CVV 1234", validator.validateCvv("1234", "visa"));
        assertFalse("Empty CVV", validator.validateCvv("", "visa"));
        assertFalse("Unknown brand", validator.validateCvv("123", "unknown"));
    }
    
    @Test
    public void testGetBrandInfo() {
        BrandData.Brand visaInfo = validator.getBrandInfo("visa");
        assertNotNull("Visa info exists", visaInfo);
        assertEquals("Visa name", "visa", visaInfo.name);
        assertNotNull("Visa regexp", visaInfo.regexpFull);
        
        assertNull("Unknown brand", validator.getBrandInfo("unknown"));
    }
    
    @Test
    public void testGetBrandInfoDetailed() {
        BrandDataDetailed.BrandDetailed visaDetailed = validator.getBrandInfoDetailed("visa");
        assertNotNull("Visa detailed exists", visaDetailed);
        assertEquals("Visa scheme", "visa", visaDetailed.scheme);
        assertEquals("Visa brand name", "Visa", visaDetailed.brand);
        assertTrue("Visa number lengths", visaDetailed.number.lengths.length > 0);
        
        assertNull("Unknown scheme", validator.getBrandInfoDetailed("unknown"));
    }
    
    @Test
    public void testFindBrandDetailed() {
        BrandDataDetailed.BrandDetailed brand = validator.findBrandDetailed("4012001037141112");
        assertNotNull("Brand detailed exists", brand);
        assertEquals("Visa scheme", "visa", brand.scheme);
        assertEquals("Visa brand name", "Visa", brand.brand);
    }
    
    @Test
    public void testListBrands() {
        String[] brands = validator.listBrands();
        assertNotNull("Brands list exists", brands);
        assertTrue("Brands list not empty", brands.length > 0);
        
        boolean hasVisa = false;
        boolean hasMastercard = false;
        for (String brand : brands) {
            if ("visa".equals(brand)) hasVisa = true;
            if ("mastercard".equals(brand)) hasMastercard = true;
        }
        
        assertTrue("Has Visa", hasVisa);
        assertTrue("Has Mastercard", hasMastercard);
    }
}
